Added options to the Prepare > Data Frame > Rename Column dialogue

N-thony opened the pull request:
Fixes #6877 
This is still a work in progress. @shadrackkibet I have added the grid under the multiple options as suggested by @rdstern in the corresponding issue. At this stage, I would like your help on the backend side. I'm not sure about how the R code or proper package to use for renaming multiple variables through the grid.

shadrackkibet commented:
@N-thony thanks for the discussion. Leave this to me for now. I will alert you when the R code is ready.

rdstern reviewed (changes requested):
@N-thony so far I have just looked at the multiple.  Not yet the rename_with.
a) Looks fine on manual changes.
b) When I paste from Excel they appear to change, but it doesn't register.
c) When I press Ok and have made a change, but not clicked elsewhere, then the last change doesn't register.
d) When I click to include labels it goes blank and freezes.  (And, once it is working please test copying the old names and pasting into the variable labels.

N-thony commented:
> @N-thony so far I have just looked at the multiple. Not yet the rename_with. a) Looks fine on manual changes. b) When I paste from Excel they appear to change, but it doesn't register. c) When I press Ok and have made a change, but not clicked elsewhere, then the last change doesn't register. d) When I click to include labels it goes blank and freezes. (And, once it is working please test copying the old names and pasting into the variable labels.

@rdstern thanks for your comment. During our discussion we agreed to get the multiple option working well before moving to the rename_with option. Thanks for reminding about the pasting feature on the grid. I made some change regarding your comments above. I didn't yet finish to make a proper test to check if it is stable, but you may wish to test it too at this stage.

rdstern reviewed (changes requested):
@N-thony that's progress!  
a) Paste is working, but only for the first cell thaqt is pasted.
b) Adding labels is now sensible.  I only checked pasting, and that seems the same as the names, so it just pastes the first label.

N-thony commented:
@rdstern I'm still looking into this. Meanwhile, @shadrackkibet I found that with the functions parameters passing the `columns `and `indexes` is a bit tricky. Using the survey dataset, I have the selected the three first column Village, field and size then through my dialogue I renamed both the name and label of village which the index is 1. Then click Ok. All runs fine, when I go back I expected to see only 2 columns here field and size since the new village name is not in the defined filter then, I do rename field column in the grid then I found it's still rename now the new village name column  since now in the grid the index of field is 1. I can demonstrate this if you have a minute for a quick call.

N-thony commented:
> @N-thony I am not sure I understand what you find tricky. I tried the following;
> 
> * Loaded Survey data
> * Prepare>Data Frame>Rename columns
> * Data Options>Columns>Define New Selection>Selected village,field,size>Add condition>Return>Return
> * Selected Multiple at the top
> * Checked Include Variable Labels checkbox
> * Renamed columns as follows & added new labels (see below);
>   ![image](https://user-images.githubusercontent.com/26301973/153915614-f0be7cea-10f5-43bc-8cf4-f6e38dfcfe0b.png)
> * Click Ok
> * I got the following code (Which is correct)
> 
> ```
> # Code generated by the dialog, Rename Columns
> data_book$rename_column_in_data(data_name="survey", type="multiple", .fn=tolower, new_column_names_df=data.frame(cols=c("village1","field2","size3"), index=c(1,2,3)), new_labels_df=data.frame(cols=c("Village","Field","Size"), index=c(1,2,3)))
> ```
> 
> Output in the Data View and Column Metadata is as follows; ![image](https://user-images.githubusercontent.com/26301973/153916084-4c8a8109-3a75-40fc-ace1-6522ac03538d.png) ![image](https://user-images.githubusercontent.com/26301973/153916148-032484d7-c39f-488d-9eb2-4ca9821df5b3.png)
> 
> From my analysis above I do not see any problem with the R code. There are definitely efficiency issues at the dialog level.
> 
> By the way, remove `.fn=tolower` from code when multiple and single options are selected.

@shadrackkibet thanks for reproducing that. I meant if you only rename the first row as below, it is fine.
![image](https://user-images.githubusercontent.com/68591383/153917824-225a42ce-fea8-455b-a15b-af4baf0bf2c7.png)

Then back to the dialog, you will have the following column present in the grid since the renamed column is not in the selection condition. Then try to rename them, `field` and `size`
![image](https://user-images.githubusercontent.com/68591383/153918017-f028c646-dcd5-4910-9d60-d7707a1624de.png)

N-thony commented:
I meant by tricky since the indexes are supposed to come from the data frame?

rdstern reviewed (changes requested):
@N-thony did you want me to check again now - well here is what I found mon multiple.
a) It automatically used all the variables.  With the usual data selector I assumed you would start with it empty and then choose which variables to rename?  
b) If I manually edit, then it seems fine.  That is with the exception that I have to exit the last change, or it is not used.  I suggest either this change is closed (and used) when you press another key.  Alternatively (not quite so good) Ok and To Script) could be disabled when you are changing a name.
c) Paste still doesn't work, when you are pasting multiple cells.  I tried both pasting new names from Excel and also copying names into labels within the dialogue.  Just the first change is implemented, e.g. just x1, if I paste x1 to x7 in one go.

rdstern reviewed (changes requested):
@N-thony that's progress, but there is quite a lot to say.
a) Possibly change the title from New Label to Label.
b) I pasted into the labels, but OK was not enabled, until I changed a name.
c) It pasted correctly.  But when I returned, those labels did not appear in your grid.
d) I was not able to paste into the new names, largely because of that very annoying message.  I suggest it is turned off.  
e) So I entered names manually aqnd ignored the message which came repeatedly.  I deliberately now has a repeated name, and pressed Ok and it crashed.  I suggest you don't do any checks when the new names are being done, but check when Ok is pressed. If the set of new names are not valid, then decide on an action.  Probably don't change them, or perhaps give an option to tidy, i.e. make valid and unique.  (I am sure we must have the code for that when importing a data file?

rdstern reviewed (changes requested):
This is looking good.  You may need help mon the R code for the rename_with.  I wonder if @HawardKetoyoMsatsi can help.  It may have to be @shadrackkibet , but Haward is pretty good with R.
On 2 details, with the Abbreviation, the up-down can be less wide.  It will never need more than 2 numbers.  The default is 8 and the minimum is 1 for now.  (It may become 2).  
And after the up/down, put characters.  
Then, not only on this dialogue, and perhaps a separate pull request, could you look carefully at the details of the regex keyboard.  The cursor is not always in the right place, after each key.

N-thony commented:
@rdstern I and @shadrackkibet agreed that the changes on `Rename With` #6877 will be done in the separate PR. Here are just the changes concerning the `Multiple` option.
@shadrackkibet I would like to be able to also have the label names in the Label on filling the grid, I'm not sure there is an R code doing that. I have done with the one returning a single label name and made a list of that on the front end side but this makes the grid not fill faster. Do you have any suggestion on that?

shadrackkibet commented:
What about `data_book$get_column_labels()`? Did you try it?

shadrackkibet commented on instat/dlgName.vb:
```diff
@@ -199,12 +393,62 @@ Public Class dlgName
+        'This is to be done in R to avoid the app to slow down when getting the list of label
+        For i As Integer = 0 To dctNameRowsValues.Count - 1
+            clsColmnLabelsRFunction.AddParameter("columns", Chr(34) & dctNameRowsValues(i) & Chr(34), iPosition:=i)
+            expItems = frmMain.clsRLink.RunInternalScriptGetValue(clsColmnLabelsRFunction.ToScript(), bSilent:=True)
```
Instead of going to R in each loop why don't you just get labels once? Then populate the grid with the labels.

N-thony commented:
> What about `data_book$get_column_labels()`? Did you try it?

Sounds good, let me try it then.

N-thony commented:
> > What about `data_book$get_column_labels()`? Did you try it?
> 
> Sounds good, let me try it then.

@shadrackkibet could check this and probably add the check for uniqueness of column names?

N-thony commented:
@shadrackkibet how about this? Any update? Thanks.

shadrackkibet commented:
@N-thony the check for duplicate column names is now added. There are "things" you need to stop a user from typing at the grid level. 

- Numeric values
- Booleans - `TRUE, FALSE, T, F`
- Column names with spaces
- Ok should not be enabled if a cell is empty.

If any of these is added to the grid just display a message box when the user is about to click OK.

N-thony commented:
@rdstern with the update in R code function by @shadrackkibet and some small bug fixes by doing the following test using the experimental survey dataset

- Pasted both `New name` and `Label` columns at the same time it works **OK**
- Renamed all cells in both `New name` and `Label` columns at the same time it works **OK**
- Renamed single cell in `New name` column it works **OK**
- Renamed single cell in `Label` column it works **OK**
I also added the additional check as suggested by Shadrack.

Note that the renamed value is considered in the R code only after leaving the cells, I have tried to fix that but seems to be a particular problem with Reogrid. I will keep looking into it. @rdstern could you test too? Thanks.

rdstern commented:
@N-thony this does seem to be causing you a lot of problems.
a) I tried with the survey data, pasting x1,... x7 from Excel into the names.  It works!.  But when I return to the dialogue the survey is still there with all the old names.  If I move data frames and then come back, then it updates.  
b) I added a data frame from DAAG called rockart.  It has 641 variables.  It takes over 20 seconds to load into the grid.  And you can't stop it.  There will be wider ones still.  So we do need an Add button, even if that is another issue.  And you should liaise with Patrick, who is looking at efficiencies here.
c) I then tried pasting a set longer than the number of names.  Good that worked fine.  It ignored me.
d) I then tried making non-unique names.  It gives an error, that is reasonable.  (Ideally this is captured earleir so we give our own message, but this is fine.)
e) then I tried a name with an error, so x<space>4.  It crashed.  These are obvious things to check for.  We must have routines in our file import that check this.
f) I tried labels and it seems to add them ok.  But then I tried to delete them and change them.  Once I have labels they don't ever seem to go away, and they don't change as easily as labels as they do in the grid.
You seem to be making quite "heavy weather" of all this.  
I was also hoping you might have made more progress on the rename with as well?

N-thony commented:
@rdstern I have fixed a couple of bugs found during our last chat, also from your review,

- especially heavy datasets like `DAAG rockart`, 
- deleting all column or single or multiple no neighborhood Label cells with Delete Keyboard
- Also fixed some bugs on testing.

lloyddewit commented:
@HawardKetoyoMsatsi Please could you review? thanks

rdstern reviewed (changes requested):
@N-thony this is much better now.  I first tried a very wide data set.  This was with over 7000 variables.  It took 16 seconds to load into the ordinary data selector, and 11 seconds to load into this grid.  Ideally both may be improved further, but we could live with this.  I also tried:
a) Pasting names - ok
b) Updating names - ok
c) Adding, updating and deleting labels - ok
d) Names repeated - sort of ok.  Told me it wasn't allowed as an error.  But then I could continue.
e) The one thing - for me so far that needs to be fixed is that it crashes if you input an illegal name, e.g. x<space>1.  We must already have a routine for this.

N-thony commented:
@rdstern I resolved item e) and also added the check for the French accent in the name for example `variety` is `variété` and it is an invalid column name. Please test it. Thanks.

rdstern reviewed (changes requested):
@N-thony I still managed to break it.  I added a single variable with an illgal name and it traps it.
Then I pasted
y.1
y_2
y-3
y4
y5
and then pressed Ok.  It crashed.

rdstern commented:
@N-thony I still get the same error, namely with:

![image](https://user-images.githubusercontent.com/11226469/156404034-e72533a8-28d5-416e-8dfe-87548d61fd6d.png)

I get:

![image](https://user-images.githubusercontent.com/11226469/156403910-edbffe7d-efd5-49b5-a732-16e2ea2192a5.png)

Then I am thrown out of R-Instat

rdstern reviewed (dismissed):
@N-thony great it now seems to work fine.  I was also keen that you removed the message about individual names being wrong.  I am happy that you did that.
I hope that @lloyddewit or @shadrackkibet can check quickly, so it can be merged.  It is an important improvement, and I hope the rename_with (to be done later) will be a relatively small task

lloyddewit commented:
> @N-thony great it now seems to work fine. I was also keen that you removed the message about individual names being wrong. I am happy that you did that. I hope that @lloyddewit or @shadrackkibet can check quickly, so it can be merged. It is an important improvement, and I hope the rename_with (to be done later) will be a relatively small task

@HawardKetoyoMsatsi I think it makes sense for you to complete your peer review first, please let us know when it's done, thanks

HawardKetoyoMsatsi reviewed (commented):
@N-thony I find the codes to be fine and running fine. I have no issue  data_object_R6.R but am having a very small issue (ignore if possible) with dlgName.vb that I have left as a comment rather than a request for change.

N-thony commented:
@rdstern @shadrackkibet I made the change as we discussed few minutes ago. Please could you check/test? Thanks.

rdstern reviewed (dismissed):
@N-thony looks fine to me.  I am approving, so it can be merged.  The next item is the abbreviate function, which I hope will be easy to add?

shadrackkibet commented:
@N-thony `abbreviate` can be another radio button. Implement as follows;

`data_book$rename_column_in_data(data_name="survey", type="rename_with", .fn=abbreviate)`

If a control (Nud/input) is needed for `minlength`(R default =4) or any other parameter can be added as follows;

`data_book$rename_column_in_data(data_name="survey", type="rename_with", .fn=abbreviate, minlength = 4)`

rdstern commented:
@shadrackkibet great to have this and a really good "quick win" for R-Instat.  Once that is added and merged, then I suggest that the last feature in renaming (which overlaps with all your select stuff, i.e. your select is the find, and then we add replace!) becomes far less urgent, and could wait till later in the year.  Certainly until you have everything you need in the Select stuff.  So, I can write that as a new issue.  (Maybe we don't even need the find, but it just works on a defined select!

shadrackkibet commented:
> @shadrackkibet great to have this and a really good "quick win" for R-Instat. Once that is added and merged, then I suggest that the last feature in renaming (which overlaps with all your select stuff, i.e. your select is the find, and then we add replace!) becomes far less urgent, and could wait till later in the year. Certainly until you have everything you need in the Select stuff. So, I can write that as a new issue. (Maybe we don't even need the find, but it just works on a defined select!

This should be straightforward to implement. In a nutshell, this will simply have a receiver to take in a select object then we just rename those columns in the select object.

rdstern commented:
Exactly!

N-thony commented:
> @N-thony `abbreviate` can be another radio button. Implement as follows;
> 
> `data_book$rename_column_in_data(data_name="survey", type="rename_with", .fn=abbreviate)`
> 
> If a control (Nud/input) is needed for `minlength`(R default =4) or any other parameter can be added as follows;
> 
> `data_book$rename_column_in_data(data_name="survey", type="rename_with", .fn=abbreviate, minlength = 4)`

@rdstern @shadrackkibet I have added this.

shadrackkibet commented on instat/dlgName.vb:
```diff
@@ -109,33 +127,335 @@ Public Class dlgName
+        dctRowsNewNameChanged.Clear()
+        dctRowsNewLabelChanged.Clear()
+
+        ucrNudAbbreviate.SetText(4)
```
If you have the control passing the parameter then the control will read from the R code.

shadrackkibet commented on instat/dlgName.vb:
```diff
@@ -16,14 +16,27 @@
     Private clsDefaultRFunction As New RFunction
+    Private clsNewColNameDataframeFunction As New RFunction
+    Private clsNewLabelDataframeFunction As New RFunction
+    Private clsDummyFunction As New RFunction
```
Try using conditions. Then you wouldn't need a dummy function. Happy to discuss.

shadrackkibet commented on instat/dlgName.vb:
```diff
@@ -109,33 +127,335 @@ Public Class dlgName
+        End If
+    End Sub
+
+    Private Function GetValuesAsVector(dctValues As Dictionary(Of Integer, String)) As String
+        Dim strValue As String = ""
+        Dim i As Integer
+        strValue = strValue & "c("
+        For Each iRow As Integer In dctValues.Keys
+            If i > 0 Then
+                strValue = strValue & ","
+            End If
+            strValue = strValue & Chr(34) & dctValues(iRow) & Chr(34)
+            i = i + 1
+        Next
+        strValue = strValue & ")"
+        Return strValue
+    End Function
+
+    Private Sub ValidateNamesFromDictionary(iColIndex As Integer)
+        If iColIndex = 1 Then
+            For Each value In dctRowsNewNameChanged.Values
+                If Not CheckNames(value, iColIndex) Then
+                    MsgBox("The column name must not be a numeric or contains space or french accent or be a boolean e.g TRUE, FALSE, T, F.")
+                    bCurrentCell = False
+                    Exit For
```
We need to be moving this kind of code into user control. I think we have this kind of code in climatic data entry. I suggest you add a TODO comment.

shadrackkibet commented on instat/dlgName.vb:
```diff
@@ -109,33 +127,335 @@ Public Class dlgName
+    End Sub
+
+    Private Sub RemoveLabelsParams()
+        clsDefaultRFunction.RemoveParameterByPosition(8)
+        clsDefaultRFunction.RemoveParameterByPosition(9)
```
In my opinion, `RemoveParameterByPosition` is difficult to maintain by developers. If a developer adds another parameter then the order isn't always "correct". I would prefer `RemoveParameterByName()`. Does this make sense to you?

N-thony replied:
this is resolved

rdstern reviewed (changes requested):
@N-thony I think the abbreviate works fine now.  But your function has now (behind the scenes) changed the janitor function for clean names.  I think it also applies the default width from the abbreviate function.  
At least the result of the clean names, with snake works fine in version 0.7.4 and is now different in your current function.  I was trying with the iris set.

rdstern reviewed (dismissed):
Great - this is now an excellent new feature.  Hope @shadrackkibet or @lloyddewit can approve, so it can be merged

lloyddewit commented on instat/dlgName.vb:
```diff
@@ -213,7 +551,7 @@ Public Class dlgName
     End Sub
 
-    Private Sub ucrCoreControls_ControlContentsChanged() Handles ucrInputNewName.ControlContentsChanged, ucrReceiverName.ControlContentsChanged, ucrSelectVariables.ControlContentsChanged, ucrPnlOptions.ControlContentsChanged
+    Private Sub ucrCoreControls_ControlContentsChanged() Handles ucrInputNewName.ControlContentsChanged, ucrReceiverName.ControlContentsChanged, ucrReceiverColumns.ControlContentsChanged, ucrSelectVariables.ControlContentsChanged, ucrPnlOptions.ControlContentsChanged
```
@HawardKetoyoMsatsi thanks for the comment, I agree that this may not be strictly necessary but is safe enough

lloyddewit commented:
@N-thony I think there are 3 open comments from @shadrackkibet that are still open. Please could you fix these (without setting to resolved) and then ask @shadrackkibet to check. If he's happy, then he can resolve these remaining comments and approve.

@shadrackkibet thanks for reviewing this PR, I don't think I need to review also, but if you'd like me to review, then let me know, thanks

N-thony commented:
> @N-thony I think there are 3 open comments from @shadrackkibet that are still open. Please could you fix these (without setting to resolved) and then ask @shadrackkibet to check. If he's happy, then he can resolve these remaining comments and approve.
> 
> @shadrackkibet thanks for reviewing this PR, I don't think I need to review also, but if you'd like me to review, then let me know, thanks

@lloyddewit I think the three comments were already solved.

lloyddewit commented:
> 
> @lloyddewit I think the three comments were already solved.

@N-thony I looked at the code and couldn't see where the comments are solved. Maybe I'm missing something. For each comment, please could you explain where/how the comment is solved?
thanks

lloyddewit commented:
@N-thony thank you for the update
@shadrackkibet if you think that all your comments have been resolved, then please could you approve?
Thanks
